grr/server/grr_response_server/gui/ui/lib/queued_exhaust_map_test.ts

```diff
@@ -0,0 +1,88 @@
+import { of, from, Observable } from 'rxjs';
+import { queuedExhaustMap } from './queued_exhaust_map';
+import { delay } from 'rxjs/operators';
+import { fakeAsync, tick } from '@angular/core/testing';
+
+describe('queuedExhaustMap', () => {
+  it('shouldn\'t discard anything if there is just one element (queue size 1)', () => {
+    const valuePassed = 'dummy value';
+
+    const singleValue$ = of(valuePassed).pipe(
+      queuedExhaustMap((value) => {
+        return of(value);
+      }, 1),
+    );
+
+    const processedValues = getProcessedValues(singleValue$);
+    expect(processedValues).toEqual([valuePassed]);
+  });
+
+  it('shouldn\'t discard anything if there are no concurrent elements (queue size 1)', () => {
+    const repeatTimes = 10;
+    const sentValues = Array(repeatTimes).map((_value, index) => index);
+
+    const manyValues$ = from(sentValues).pipe(
+      queuedExhaustMap((value) => {
+        return of(value);
+      }, 1),
+    );
+
+    const processedValues = getProcessedValues(manyValues$);
+    expect(processedValues).toEqual(sentValues);
+  });
+
+  it('should discard everything but the last concurrent element (queue size 1)', fakeAsync(() => {
+    const blockTimeMs = 1000;
+
+    const manyValuesWithDelay$ = of('initial', 'discarded1', 'discarded2', 'lastSaved').pipe(
+      queuedExhaustMap((value) => {
+        return of(value).pipe(
+          delay(blockTimeMs),
+        );
+      }, 1),
+    );
+
+    const processedValues = getProcessedValues(manyValuesWithDelay$);
+
+    tick(blockTimeMs);
+    expect(processedValues).toEqual(['initial']);
```

Comment:
Here it gets a bit confusing. There's an assumption that processedValues will change behind the scenes, since its "captured" in a closure inside getProcessedValues. That's maybe ok for tests, but is somehow hard to follow.

What about changing the interface of getProcessedValues() to:
```
getProcessedValues<T>(observable$: Observable<T>): () => ReadonlyArrat<T>
```

Basically, getProcessedValues() will return a callable. This callable will be returning all the emitted elements from the last call.

Then your code here would look like:
```
const processedValuesFn = getProcessedValues(manyValuesWithDelay$);

tick(blockTimeMs);
expect(processedValuesFn()).toEqual(['initial']);

tick(blockTimeMs);
expect(processedValuesFn()).toEqual(['lastSaved']);
```

Then what previously was happening behind-the-scenes is now explicit and connected to a function call. What do you think?